xds/internal/client/filter_chain.go

```diff
@@ -0,0 +1,570 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package client
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	v3listenerpb "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	v3tlspb "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+	"github.com/golang/protobuf/proto"
+
+	"google.golang.org/grpc/xds/internal/version"
+)
+
+// Represents a wildcard IP prefix. Go stdlib `Contains()` method works for both
+// v4 and v6 addresses when used on this wildcard address.
+const emptyAddrMapKey = "0.0.0.0/0"
+
+var (
+	// Parsed wildcard IP prefix.
+	_, zeroIP, _ = net.ParseCIDR("0.0.0.0/0")
+)
+
+// FilterChain captures information from within a FilterChain message in a
+// Listener resource.
+//
+// Currently, this simply contains the security configuration found in the
+// 'transport_socket' field of the filter chain. The actual set of filters
+// associated with this filter chain are not captured here, since we do not
+// support these filters on the server-side yet.
+type FilterChain struct {
+	// SecurityCfg contains transport socket security configuration.
+	SecurityCfg *SecurityConfig
+}
+
+// SourceType specifies the connection source IP match type.
+type SourceType int
+
+const (
+	// SourceTypeAny matches connection attempts from any source.
+	SourceTypeAny SourceType = iota
+	// SourceTypeSameOrLoopback matches connection attempts from the same host.
+	SourceTypeSameOrLoopback
+	// SourceTypeExternal matches connection attempts from a different host.
+	SourceTypeExternal
+)
+
+// FilterChainManager contains all the match criteria specified through all
+// filter chains in a single Listener resource. It also contains the default
+// filter chain specified in the Listener resource. It provides two important
+// pieces of functionality:
+// 1. Validate the filter chains in an incoming Listener resource to make sure
+//    that there aren't filter chains which contain the same match criteria.
+// 2. As part of performing the above validation, it builds an internal data
+//    structure which will if used to look up the matching filter chain at
+//    connection time.
+//
+// The logic specified in the documentation around the xDS FilterChainMatch
+// proto mentions 8 criteria to match on. gRPC does not support 4 of those, and
+// we get rid of filter chains which contain any of these unsupported fields at
```

Comment:
"get rid of"? Ignore? NACK?

Reply:
Replaced with `ignore` and added a TODO to capture the fact that these should not be ignored at validation time, but only at connection time.